`withNumberCompressed(true)` silently corrupts boxed `Long` fields when serialize and deserialize go through different `ThreadSafeForyPool` instances

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

- Apache Fory 0.17.0
- JDK 21
- Linux / macOS

### Component(s)

Java

### Minimal reproduce step

When **all** of the following are true, Fory either throws `IndexOutOfBoundsException` from the generated codec or, worse, silently decodes a different long value with no exception:

- pooled Fory (`buildThreadSafeForyPool`)
- codegen on (`withCodegen(true)` — the default)
- number compression on (`withNumberCompressed(true)`)
- a field of type boxed `Long` (not primitive `long`)

The bug requires two pool instances. Single-Fory round-trip works. Disabling any one of `withNumberCompressed`, `withCodegen`, the boxing, or the second pool eliminates the bug.

```java
import org.apache.fory.Fory;
import org.apache.fory.ThreadSafeFory;
import org.apache.fory.config.CompatibleMode;
import org.apache.fory.config.ForyBuilder;
import org.apache.fory.config.Language;

public class ForyPoolNumberCompressedBug {

    public record Payload(Long longValue, String stringValue) {}

    static ForyBuilder builder() {
        return Fory.builder()
                .withLanguage(Language.JAVA)
                .withCodegen(true)
                .withAsyncCompilation(true)
                .requireClassRegistration(false)
                .suppressClassRegistrationWarnings(true)
                .withDeserializeUnknownClass(true)
                .withRefTracking(true)
                .withCompatibleMode(CompatibleMode.COMPATIBLE)
                .withStringCompressed(true)
                .withNumberCompressed(true) // <-- removing this fixes the bug
                .withRefCopy(true);
    }

    public static void main(String[] args) {
        // Two pools built from the same config — e.g. producer/consumer on
        // different nodes, or process-restart snapshot restore.
        ThreadSafeFory writer = builder().buildThreadSafeForyPool(4);
        ThreadSafeFory reader = builder().buildThreadSafeForyPool(4);

        Payload original = new Payload(
                123_456_789L,
                "longer string with multibyte: \u00ff\u00fe");

        byte[] bytes = writer.serialize(original);
        Payload roundTrip = (Payload) reader.deserialize(bytes);

        System.out.println("original  = " + original);
        System.out.println("roundTrip = " + roundTrip);

        if (!original.equals(roundTrip)) {
            throw new AssertionError("CORRUPTION: " + original + " != " + roundTrip);
        }
    }
}
```

### What did you expect to see?

A round-trip across two pools built from the same `ForyBuilder` should produce a value equal to the original.


### What did you see instead?


```
original  = Payload[longValue=123456789, stringValue=longer string with multibyte: ÿþ]
roundTrip = Payload[longValue=988764757, stringValue=longer string with multibyte: ÿþ]
Exception in thread "main" java.lang.AssertionError: CORRUPTION: ...
```

Alternate failure mode (thrown from generated codec)

With other payload shapes (concurrent calls on a single pool, multiple records in sequence) the same configuration throws:

```
java.lang.IndexOutOfBoundsException: readerIndex(78) + length(18) exceeds size(84)
    at org.apache.fory.memory.MemoryBuffer$BoundChecker.fillBuffer(MemoryBuffer.java:189)
    at org.apache.fory.serializer.StringSerializer.readBytesUnCompressedUTF16(StringSerializer.java:565)
    at org.apache.fory.serializer.StringSerializer.readCompressedBytesString(StringSerializer.java:259)
    at <Pkg>$PayloadForyRefCodecMetaShared0_0.read(<Pkg>$PayloadForyRefCodecMetaShared0_0.java:47)
    at org.apache.fory.context.ReadContext.readDataInternal(ReadContext.java:666)
    at org.apache.fory.context.ReadContext.readNonRef(ReadContext.java:580)
    at org.apache.fory.context.ReadContext.readRef(ReadContext.java:518)
    at org.apache.fory.Fory.deserialize(Fory.java:476)
```

The generated codec name is `*ForyRefCodecMetaShared0_0`, suggesting the writer and reader codecs disagree on the encoded layout for boxed numeric fields.

### Bisection (each toggle applied in isolation against the failing baseline)

| Toggle | Result |
|---|---|
| `withNumberCompressed(false)` | works |
| `withCodegen(false)` (interpreted path) | works |
| `Payload(long longValue, String stringValue)` (primitive long) | works |
| single Fory (no pool) | works |
| `withStringCompressed(false)` | **still broken** |
| `withRefTracking(false)` + `withRefCopy(false)` | **still broken** |
| `CompatibleMode.SCHEMA_CONSISTENT` | **still broken** |

So the trigger is codegen + number compression + boxed numeric field + cross-pool deserialization. The bug exists in both `COMPATIBLE` and `SCHEMA_CONSISTENT` modes.

### Anything Else?

As a workaround, resorting to `withNumberCompressed(false)`

--- a/java/fory-core/src/main/java/org/apache/fory/builder/BaseObjectCodecBuilder.java
+++ b/java/fory-core/src/main/java/org/apache/fory/builder/BaseObjectCodecBuilder.java
@@ -1856,17 +1856,15 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
     } else {
       if (typeRef.isPrimitive() && !nullable) {
         // Only skip null check if BOTH: local type is primitive AND sender didn't write null flag
-        Expression value = deserializeForNotNull(buffer, typeRef, null);
+        Expression value = deserializeForNotNullForField(buffer, descriptor, null);
         // Should put value expr ahead to avoid generated code in wrong scope.
         return new ListExpression(value, callback.apply(value));
       }
-      // Pass local field type so readNullable can use default value for primitives when null
-      Class<?> localFieldType = typeRef.isPrimitive() ? typeRef.getRawType() : null;
       return readNullableField(
           buffer,
           descriptor,
           callback,
-          () -> deserializeForNotNull(buffer, typeRef, null),
+          () -> deserializeForNotNullForField(buffer, descriptor, null),
           nullable);
     }
   }

--- a/java/fory-core/src/main/java/org/apache/fory/codegen/Expression.java
+++ b/java/fory-core/src/main/java/org/apache/fory/codegen/Expression.java
@@ -56,6 +56,7 @@ import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
+import java.util.Locale;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import org.apache.fory.memory.Platform;
@@ -430,7 +431,8 @@ public interface Expression {
             return new ExprCode(
                 FalseLiteral, new LiteralValue(javaType, "Float.NEGATIVE_INFINITY"));
           } else {
-            return new ExprCode(FalseLiteral, new LiteralValue(javaType, String.format("%fF", f)));
+            return new ExprCode(
+                FalseLiteral, new LiteralValue(javaType, String.format(Locale.ROOT, "%fF", f)));
           }
         } else if (javaType == Double.class) {
           Double d = (Double) value;
@@ -443,7 +445,8 @@ public interface Expression {
             return new ExprCode(
                 FalseLiteral, new LiteralValue(javaType, "Double.NEGATIVE_INFINITY"));
           } else {
-            return new ExprCode(FalseLiteral, new LiteralValue(javaType, String.format("%fD", d)));
+            return new ExprCode(
+                FalseLiteral, new LiteralValue(javaType, String.format(Locale.ROOT, "%fD", d)));
           }
         } else if (javaType == Byte.class) {
           return new ExprCode(

--- a/java/fory-latest-jdk-tests/src/test/java/org/apache/fory/integration_tests/RecordSerializersTest.java
+++ b/java/fory-latest-jdk-tests/src/test/java/org/apache/fory/integration_tests/RecordSerializersTest.java
@@ -28,7 +28,10 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
 import org.apache.fory.Fory;
+import org.apache.fory.ThreadSafeFory;
 import org.apache.fory.config.CompatibleMode;
+import org.apache.fory.config.ForyBuilder;
+import org.apache.fory.config.Language;
 import org.apache.fory.context.MetaReadContext;
 import org.apache.fory.context.MetaWriteContext;
 import org.apache.fory.test.bean.Struct;
@@ -41,6 +44,8 @@ import org.testng.annotations.Test;
 public class RecordSerializersTest {
 
   public record Foo(int f1, String f2, List<String> f3, char f4) {}
+
+  public record NumberCompressedPayload(Long longValue, String stringValue) {}
 
   @Test
   public void testIsRecord() {
@@ -78,6 +83,37 @@ public class RecordSerializersTest {
     Fory fory = Fory.builder().requireClassRegistration(false).withCodegen(codegen).build();
     Foo foo = new Foo(10, "abc", new ArrayList<>(Arrays.asList("a", "b")), 'x');
     Assert.assertEquals(fory.deserialize(fory.serialize(foo)), foo);
+  }
+
+  @Test
+  public void testNumberCompressedBoxedLongRecordRoundTripAcrossPools() {
+    ThreadSafeFory writer = newNumberCompressedPool();
+    ThreadSafeFory reader = newNumberCompressedPool();
+
+    NumberCompressedPayload payload =
+        new NumberCompressedPayload(123_456_789L, "longer string with multibyte: \u00ff\u00fe");
+
+    byte[] bytes = writer.serialize(payload);
+    Assert.assertEquals(reader.deserialize(bytes), payload);
+  }
+
+  private static ThreadSafeFory newNumberCompressedPool() {
+    return newNumberCompressedBuilder().buildThreadSafeForyPool(4);
+  }
+
+  private static ForyBuilder newNumberCompressedBuilder() {
+    return Fory.builder()
+        .withLanguage(Language.JAVA)
+        .withCodegen(true)
+        .withAsyncCompilation(false)
+        .requireClassRegistration(false)
+        .suppressClassRegistrationWarnings(true)
+        .withDeserializeUnknownClass(true)
+        .withRefTracking(true)
+        .withCompatibleMode(CompatibleMode.COMPATIBLE)
+        .withStringCompressed(true)
+        .withNumberCompressed(true)
+        .withRefCopy(true);
   }
 
   @Test(dataProvider = "codegen")

--- a/java/fory-latest-jdk-tests/src/test/java/org/apache/fory/integration_tests/RecordSerializersTest.java
+++ b/java/fory-latest-jdk-tests/src/test/java/org/apache/fory/integration_tests/RecordSerializersTest.java
@@ -22,6 +22,7 @@ package org.apache.fory.integration_tests;
 import static org.apache.fory.collection.Collections.ofArrayList;
 import static org.apache.fory.collection.Collections.ofHashMap;
 
+import java.io.Serializable;
 import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
 import java.util.ArrayList;
@@ -46,6 +47,9 @@ public class RecordSerializersTest {
   public record Foo(int f1, String f2, List<String> f3, char f4) {}
 
   public record NumberCompressedPayload(Long longValue, String stringValue) {}
+
+  public record BoxedPrimitiveRecord(String lensId, Long from, Long to, Integer type)
+      implements Serializable {}
 
   @Test
   public void testIsRecord() {
@@ -95,6 +99,25 @@ public class RecordSerializersTest {
 
     byte[] bytes = writer.serialize(payload);
     Assert.assertEquals(reader.deserialize(bytes), payload);
+  }
+
+  @Test
+  public void testCompatibleCodegenBoxedPrimitiveRecordRoundTrip() {
+    Fory fory =
+        Fory.builder()
+            .withLanguage(Language.JAVA)
+            .requireClassRegistration(false)
+            .withCompatibleMode(CompatibleMode.COMPATIBLE)
+            .withClassVersionCheck(true)
+            .withCodegen(true)
+            .build();
+    BoxedPrimitiveRecord record =
+        new BoxedPrimitiveRecord(
+            "aaaaaaaa-bbbb-cccc-dddd-eeeeeeeeeeee:11111111-2222-3333-4444-555555555555",
+            123456789012345L,
+            98765432109876L,
+            146);
+    Assert.assertEquals(fory.deserialize(fory.serialize(record)), record);
   }
 
   private static ThreadSafeFory newNumberCompressedPool() {

